### PR TITLE
cache loadUser if not exists

### PR DESCRIPTION
**Pull request by butonic**
## TL;dr

This PR caches a non existing user in the DB, killing ~30 queries on every request when an ldap user logs in.

## Long version

user_ldap (type authentication) depends upon the AvatarManager, which in turn triggers a setupfs(), which in turn pulls in all apps of type 'filesystem' (activity, dav, files, files_sharing, files_trashbin and files_versions. Probably objectstore, search_elastic / search_lucene and whatever app has type filesystem).

Loading an app causes the I10n to be initialized whenever \OC::$server->getL10N() is used in the app.php directly or if it is a dependency of the other classes used when enabling the app.

L10N tries to get the user object which somewhere calls userExists() ... aaaand that is why for every request we do ~30 ` SELECT "uid", "displayname" FROM "oc_users" WHERE LOWER("uid") = LOWER(?)` queries. when loading the user_ldap app only user_database is enabled and since a non existing user is not cached the userExists check is constantly walking to the db ...

There are probably other things doing a userExists check, but L10N is the most frequent. 

@DeepDiver1975 I don't know if the user backend loading order may be the cause for the mixed UI languages issue you were investigating. It sure smells like it.

Yes, this only removes a symptom of the mess described above. Maybe this is no longer an issue whan the user table has been merged. I don't know. Looking at the app loading made my head hurt because we are constantly calling \OC_App::registerAutoloading()  for apps that don't even match the type https://github.com/owncloud/core/blob/fe1806c5a7f09a9c6cf6669cdce7986b1df9231d/lib/private/legacy/app.php#L108-L114. And then we are calling it again when loading the app: https://github.com/owncloud/core/blob/fe1806c5a7f09a9c6cf6669cdce7986b1df9231d/lib/private/legacy/app.php#L143-L143 :hankey: 

cc @PhilippSchaffrath

### CODE REVIEW

**Comment by butonic** [comment]
oh and because the user_ldap backend is not yet loaded adding this caching to the User Manager doesn't help. It'ts a mess ... @felixboehm

File: lib/private/User/Database.php [hunk 1]
```diff
@@ -78,6 +78,7 @@ public function __construct() {
 	 * itself, not in its subclasses.
 	 */
 	public function createUser($uid, $password) {
+        unset($this->cache[$uid]); // invalidate cache
```
Comment (PhilippSchaffrath) [inline lib/private/User/Database.php]: indentation + why do we need this? is it because a user could have been deteled and re-created while cache entries still exist?

File: lib/private/User/Database.php [hunk 2]
```diff
@@ -206,6 +207,7 @@ public function checkPassword($uid, $password) {
 				if(!empty($newHash)) {
 					$this->setPassword($uid, $password);
 				}
+				unset($this->cache[$uid]); // invalidate cache
```
Comment (PhilippSchaffrath) [inline lib/private/User/Database.php]: is this just so a login returns fresh data?

File: lib/private/User/Database.php [hunk 3]
```diff
@@ -229,10 +231,12 @@ private function loadUser($uid) {
...
 			}
 
-			while ($row = $result->fetchRow()) {
+			if ($row = $result->fetchRow()) {
 				$this->cache[$uid]['uid'] = $row['uid'];
 				$this->cache[$uid]['displayname'] = $row['displayname'];
-			}
+			} else {
+			    $this->cache[$uid] = false;
+            }
```
Comment (PhilippSchaffrath) [inline lib/private/User/Database.php]: indentation

File: lib/private/User/Database.php (hunk 2 above)
Comment (butonic) [inline lib/private/User/Database.php]: before checkPassword the user might have been cachad as non existing. this clears the cache and allows subsequent calls to return the new user object.

**Comment by DeepDiver1975** [comment]
Test\User\DatabaseTest::testLogin
Failed asserting that false is true.

/var/lib/jenkins/workspace/owncloud-core_core_PR-27068-JVANWQJORMVZ2ZFTRBX7EDETA7IPFGHFJDKLNVWCO3LGWDFOURRA/tests/lib/User/BackendTestCase.php:84

**Comment by PVince81** [comment]
Ok now I understand this PR. I think it makes sense.

**Comment by PVince81** [comment]
I'm wondering about other code paths affected by this cache value.

For example when deleting a DB user, the cache value should also be reset. Or it should be set to `false` directly.

**Comment by butonic** [comment]
@PVince81 when deleting a user the cache is unset in https://github.com/owncloud/core/pull/27068/files#diff-632a8e4a922463461cb1112b3cfc800bR106

**Comment by PVince81** [comment]
Ah yes indeed. It wasn't in your changeset but old code.

**Comment by guruz** [comment]
(I always wondered why oC does not cache more stuff in the PHP session. Is it because the PHP session could be much more long lived?)

**Comment by PVince81** [comment]
I'm fine with the fix then.

Would be good to have a unit test to check for the proper cache clearing.

@butonic should someone take over ?

**Comment by butonic** [comment]
@PVince81 that would bring tears of joy to my eyes ...

File: lib/private/User/Database.php [hunk 4]
```diff
@@ -220,7 +223,7 @@ public function checkPassword($uid, $password) {
 	 * @return boolean
 	 */
 	private function loadUser($uid) {
-		if (empty($this->cache[$uid])) {
+		if (!isset($this->cache[$uid]) && $this->cache[$uid] !== false) {
```
Comment (PVince81) [inline lib/private/User/Database.php]: I'll change this to `array_key_exists` here for better clarity.

Here we only want to check if the key exists even though it can have the value `false`. But `isset` would return `false` even if the key exists but its value is falsy.
`array_key_exists` will do the check we are actually expecting without confusion.

Comment (PVince81) [inline lib/private/User/Database.php]: Uh oh... `array_key_exists` doesn't work here. But it works in a separate script.

Maybe 'CappedMemoryCache` is not overriding shit properly...

I sense recursive bug finding mode approaching...

Comment (PVince81) [inline lib/private/User/Database.php]: Okay, two unhappy findings...

1. CappedMemoryCache (which implements `ArrayAccess`) uses `isset` too for key existence: https://github.com/owncloud/core/blob/master/lib/private/Cache/CappedMemoryCache.php#L41.
Since we're anyway using `isset` almost everywhere this shouldn't be a problem, hopefully...

2. When debugging, `array_key_exists` doesn't seem to call any of the methods of the capped cache, so probably not even compatible with `ArrayAccess`. And that was tested on PHP 7.0 ...

Ok, enough diverging... I'll set this back as I found it.

**Comment by PVince81** [comment]
> @PVince81 that would bring tears of joy to my eyes ...

@butonic take this napkin.

I've added unit tests. If you comment out some of the `unset` they will fail.

@jvillafanez @IljaN @DeepDiver1975 review ?

File: lib/private/User/Database.php [hunk 5]
```diff
@@ -78,7 +78,9 @@ public function __construct() {
 	 * itself, not in its subclasses.
 	 */
 	public function createUser($uid, $password) {
+		unset($this->cache[$uid]); // make sure we are reading from the db
 		if (!$this->userExists($uid)) {
+			unset($this->cache[$uid]); // invalidate non existing user in cache
```
Comment (jvillafanez) [inline lib/private/User/Database.php]: This seems unneeded if the first thing we do is unsetting the same uid.

File: lib/private/User/Database.php (hunk 2 above)
Comment (jvillafanez) [inline lib/private/User/Database.php]: I'm still unsure with this change... what if a valid user checks its right password? it shouldn't be needed to reset the cache.
What about resetting if the password is wrong?

File: lib/private/User/Database.php [hunk 6]
```diff
@@ -229,9 +233,12 @@ private function loadUser($uid) {
 				return false;
 			}
 
-			while ($row = $result->fetchRow()) {
+			if ($row = $result->fetchRow()) {
 				$this->cache[$uid]['uid'] = $row['uid'];
 				$this->cache[$uid]['displayname'] = $row['displayname'];
+			} else {
+			    $this->cache[$uid] = false;
```
Comment (jvillafanez) [inline lib/private/User/Database.php]: Indentation here

File: lib/private/User/Database.php [hunk 7]
```diff
@@ -229,9 +233,12 @@ private function loadUser($uid) {
 				return false;
 			}
 
-			while ($row = $result->fetchRow()) {
+			if ($row = $result->fetchRow()) {
```
Comment (jvillafanez) [inline lib/private/User/Database.php]: If we're relying that the query result will be unique, I'd add a comment here.

File: lib/private/User/Database.php (hunk 5 above)
Comment (PVince81) [inline lib/private/User/Database.php]: `userExists` will see that the user does not exist and it will cache `false` for that user, so we need to reset it here.

If you comment out that line some of the new tests will fail.

File: lib/private/User/Database.php (hunk 2 above)
Comment (PVince81) [inline lib/private/User/Database.php]: I think this part of the function has nothing to do with checking the password.

In old versions of OC the hashing algo used for passwords was different and less secure.
Here if we found that the password matches, we take the opportunity to rehash the password with the new algo, since the plain text password is available. This is why `setPassword` is called.

File: lib/private/User/Database.php (hunk 7 above)
Comment (PVince81) [inline lib/private/User/Database.php]: `uid` is the primary key of the table so it can only be unique. (I also wondered why `while` was used before, probably a copy-paste mistake)

File: lib/private/User/Database.php [hunk 8]
```diff
@@ -205,6 +207,7 @@ public function checkPassword($uid, $password) {
 			if(\OC::$server->getHasher()->verify($password, $storedHash, $newHash)) {
 				if(!empty($newHash)) {
 					$this->setPassword($uid, $password);
+					unset($this->cache[$uid]); // invalidate cache
```
Comment (PVince81) [inline lib/private/User/Database.php]: @jvillafanez I moved this here now to make it cleared that it's `setPassword` that might indirectly cache the user through its `userExists` call

**Comment by PVince81** [comment]
@jvillafanez https://github.com/owncloud/core/pull/27068/commits/414d3d533955fd291047d426682ed4f6b84d46ac

I also added `closeCursor` just to be safe

File: lib/private/User/Database.php (hunk 5 above)
Comment (jvillafanez) [inline lib/private/User/Database.php]: Maybe it's clearer if we unset the user if it's created correctly in the DB.

File: lib/private/User/Database.php (hunk 8 above)
Comment (jvillafanez) [inline lib/private/User/Database.php]: looks better now 👍

File: lib/private/User/Database.php (hunk 5 above)
Comment (PVince81) [inline lib/private/User/Database.php]: done

**Comment by jvillafanez** [comment]
👍 as long as the tests pass

**Comment by PVince81** [comment]
@butonic backport down to 9.1 only or further ?

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/27174

**Comment by PVince81** [comment]
stable9: https://github.com/owncloud/core/pull/27351
